Fix various registration sharing issues

ikapelyukhin opened the pull request:
* Store replay logs for deregistration messages if the request fails
* Fix logging in deleting a sibling registration subroutine

ikapelyukhin committed:
Fix various registration sharing issues

* Store replay logs for deregistration messages if the request fails
* Fix logging in deleting a sibling registration subroutine
* Replay logs in the correct (file modification time) order

ikapelyukhin commented:
@rjschwei argh, too early -- needs more fixes.

ikapelyukhin commented:
@rjschwei Currently each apache worker writes replay logs into it's own file. The problem is that when `upgrade` and `rollback` API calls happen, they might end up served by different workers.

File modification time is actually not that useful, because there are many requests in each replay file. As the result the sibling node might still get the changes in the reverse order. `flock` should prevent logs contention.

rjschwei commented on www/perl-lib/SMT/RegistrationSharing.pm:
```diff
@@ -11,12 +11,14 @@ use DBI qw(:sql_types);
-# A somewhat random number per thread avoid log and lock contention
-my $anyNumber = int(rand(1000));
+# pid of the parent apache process
+my $anyNumber = getppid();
```
This will lead to data corruption, in Perl getpid() is always the same number in any thread of an application. Each thread will write to the replay log at different times and depending on buffering data will be interleaved and thus inconsistent and corrupted.

ikapelyukhin replied:
`flock` was taking care of writing correctly, but I agree -- slurping the whole file into memory isn't a good idea. I've reverted those changes, for now at least.

ikapelyukhin committed:
Allow paths with slashes for RMT

RMT controller that handles SMT registrations will be mounted at a different path than /.

ikapelyukhin commented:
@rjschwei Let's drop the replay order changes (as it doesn't really solve the wrong order issue) and get this merged without them for now? At least it will make SMT -> RMT regsharing possible.

rjschwei commented:
@ikapelyukhin fine with me let me know when you have made those changes

ikapelyukhin committed:
Drop replay log order changes

ikapelyukhin commented:
@rjschwei Done.

felixsch reviewed (approved):
lgtm :+1: and customers should never reach this code right? Or is there a supported setup for sharing registrations between customer smt's?

ikapelyukhin commented:
@felixsch It's in `smt-ha` package:

https://github.com/SUSE/smt/blob/dfbc5d8f428200a5c4f09f8930fe220f8097e94d/package/smt.spec#L299-L303

felixsch commented:
Okay but this should also fix calling the module from apache for customers anyway. I'm fine with this changes

ikapelyukhin committed:
Merge branch 'SMT12' into registration-sharing-fixes

ikapelyukhin committed:
Bump up version; update changes file
